File: liveclaudecode/runs.py
"""Turning a directory of transcripts into one run tree.

Claude Code stores a project's transcripts as `<session>.jsonl`, with any
subagent it spawned under `<session>/subagents/<agent>.jsonl` next to an
`<agent>.meta.json`. The meta file records the `toolUseId` of the Agent call
that created it, which is what lets us rebuild the real spawn hierarchy:
the transcript containing that tool call is the parent.
"""

import json
import re
import time
from pathlib import Path

from .transcript import SCANS, get_scan, plain_text

PROJECTS = Path.home() / ".claude" / "projects"


def project_dir_for(cwd: str) -> Path:
    """Claude Code slugifies the working directory to name the project dir."""
    return PROJECTS / str(cwd).replace("/", "-")


def newest_project_dir() -> Path:
    dirs = [p for p in PROJECTS.iterdir() if p.is_dir()]
    if not dirs:
        raise SystemExit(f"no transcripts found under {PROJECTS}")
    return max(dirs, key=lambda p: p.stat().st_mtime)


_prompt_cache = {}


def first_prompt(path: Path) -> str:
    """The opening user message, used as a session's human-readable label."""
    k = str(path)
    if k in _prompt_cache:
        return _prompt_cache[k]
    text = ""
    try:
        with path.open() as fh:
            for i, line in enumerate(fh):
                if i > 60:
                    break
                try:
                    d = json.loads(line)
                except Exception:
                    continue
                if d.get("type") != "user":
                    continue
                t = plain_text((d.get("message") or {}).get("content"))
                t = re.sub(r"<command-(name|message|args)>", " ", t)
                t = re.sub(r"<[^>]+>", " ", t)
                t = re.sub(r"\s+", " ", t).strip()
                if t and not t.startswith("Caveat:"):
                    text = t[:100]
                    break
    except Exception:
        pass
    _prompt_cache[k] = text
    return text


def collect(project_dir: Path, max_age_h: float):
    """Every transcript in the project touched within `max_age_h` hours."""
    cutoff = time.time() - max_age_h * 3600
    items = []
    for jf in sorted(project_dir.glob("*.jsonl")):
        if jf.stat().st_mtime < cutoff:
            continue
        items.append({"key": jf.stem, "path": str(jf), "kind": "session",
                      "sid": jf.stem, "meta": {},
                      "label": first_prompt(jf) or jf.stem[:8]})
    for af in sorted(project_dir.rglob("subagents/*.jsonl")):
        if af.stat().st_mtime < cutoff:
            continue
        meta = {}
        mp = af.with_suffix(".meta.json")
        if mp.exists():
            try:
                meta = json.loads(mp.read_text())
            except Exception:
                pass
        sid = af.parent.parent.name
        items.append({"key": "%s/%s" % (sid, af.stem), "path": str(af),
                      "kind": "subagent", "sid": sid, "meta": meta,
                      "label": meta.get("description") or af.stem})
    return items


def build_tree(project_dir: Path, hours: float):
    """(roots, by_key) — every node carries its own stats and subtree rollups."""
    items = collect(project_dir, hours)
    by_key = {}
    for it in items:
        sc = get_scan(Path(it["path"]))
        node = dict(it)
        node.update(sc.stats())
        node["agentType"] = it["meta"].get("agentType", "") if it["kind"] == "subagent" else ""
        node["toolUseId"] = it["meta"].get("toolUseId")
        node["children"] = []
        node.pop("meta", None)
        node.pop("path", None)
        by_key[node["key"]] = node

    # Which transcript issued each Agent call, and whether it has returned.
    owner, spawn_state = {}, {}
    for it in items:
        sc = SCANS[it["path"]]
        for tid in sc.spawn_ids:
            owner[tid] = it["key"]
            spawn_state[tid] = "running" if tid in sc.open_tools else "returned"

    roots = []
    for node in by_key.values():
        node["spawnState"] = spawn_state.get(node.get("toolUseId"), "")
        pk = owner.get(node.get("toolUseId")) if node["toolUseId"] else None
        if pk and pk in by_key and pk != node["key"]:
            by_key[pk]["children"].append(node)
        elif node["kind"] == "subagent" and node["sid"] in by_key:
            # Nested agent whose spawning call we could not find: fall back to
            # hanging it off its owning session rather than dropping it.
            by_key[node["sid"]]["children"].append(node)
        else:
            roots.append(node)

    for r in roots:
        rollup(r)
    roots.sort(key=lambda n: n["subLast"] or "", reverse=True)
    return roots, by_key


def rollup(n):
    """Aggregate subtree totals onto every node (agents, tools, errors, files)."""
    agents = running = 0
    errors = n["errors"]
    tools = n["tools"]
    files = {f["path"]: f["ops"] for f in n["files"]}
    last = n["lastTs"] or ""
    live = n["live"]
    for c in n["children"]:
        rollup(c)
        agents += 1 + c["subAgents"]
        running += (1 if c["spawnState"] == "running" or c["live"] else 0) + c["subRunning"]
        errors += c["subErrors"]
        tools += c["subTools"]
        for p, o in c["subFiles"].items():
            files[p] = files.get(p, 0) + o
        last = max(last, c["subLast"] or "")
        live = live or c["subLive"]
    n["subAgents"] = agents
    n["subRunning"] = running
    n["subErrors"] = errors
    n["subTools"] = tools
    n["subFiles"] = files
    n["subLast"] = last
    n["subLive"] = live
    n["children"].sort(key=lambda c: c["firstTs"] or "")
    return n


def flatten(n, depth=0, out=None):
    """Depth-ordered lanes for the timeline view."""
    out = [] if out is None else out
    out.append({
        "key": n["key"], "label": n["label"], "agentType": n["agentType"],
        "kind": n["kind"], "depth": depth, "firstTs": n["firstTs"],
        "lastTs": n["lastTs"], "live": n["live"], "errors": n["errors"],
        "tools": n["tools"], "spawnState": n["spawnState"],
        "files": len(n["files"]),
    })
    for c in n["children"]:
        flatten(c, depth + 1, out)
    return out


def root_of(roots, key):
    """The top-level run a given transcript belongs to."""
    def hit(n):
        return n["key"] == key or any(hit(c) for c in n["children"])
    for r in roots:
        if hit(r):
            return r
    return roots[0] if roots else None


def run_phases(root, limit=16):
    """Phases announced anywhere in the run, merged in time order.

    Explicit Wave/Phase/Slice markers win outright when the run has any, so an
    orchestrator's plan is not buried under incidental bold headings from its
    workers' reports.
    """
    phases = []

    def gather(n):
        who = n["label"] if n["kind"] == "subagent" else "main"
        for m in n["milestones"]:
            phases.append({"ts": m["ts"], "title": m["title"], "who": who,
                           "strong": m.get("strong", False)})
        for c in n["children"]:
            gather(c)

    gather(root)
    if any(m["strong"] for m in phases):
        phases = [m for m in phases if m["strong"]]
    phases.sort(key=lambda m: m["ts"] or "")
    return phases[-limit:]


def strip_node(n):
    """Node without the heavy fields the client does not need."""
    return {k: v for k, v in n.items() if k not in ("children", "subFiles")}


def path_for(project_dir: Path, key: str) -> Path:
    if "/" in key:
        sid, agent = key.split("/", 1)
        return project_dir / sid / "subagents" / ("%s.jsonl" % agent)
    return project_dir / ("%s.jsonl" % key)

File: server/utils/runs.ts
import { join, resolve, sep } from 'node:path'
import { Clock, Effect, Option, Result } from 'effect'
import * as Arr from 'effect/Array'
import * as FileSystem from 'effect/FileSystem'
import { parseClaudeSubagentMetaJson, type ClaudeSubagentMeta } from '#shared/schemas/claude'
import { InvalidRunKey, PromptCache, ScanCache } from './services'
import type {
  Milestone,
  PublicRunNode,
  RunDiagnostics,
  RunNode,
  SessionEnvironment,
  TimelineLane,
} from '#shared/types/run'
import { normalizeSessionLabel } from '#shared/utils/session-label'
import { FILE_CONCURRENCY, FileDiscoveryLimiter, ignoreNotFound, isFreshFile } from './filesystem-concurrency'
import { addParseIssueCounts } from './parse-issues'
import { claudeCostSample, estimateCosts, type ClaudeCostSample } from './cost'
import {
  addUsage,
  byTimestamp,
  bySubLastDesc,
  countUnreadable,
  emptyRunDiagnostics,
  emptyUsage,
  finishRunDiagnostics,
  freshnessCutoff,
  mergeScanDiagnostics,
  visitNodes,
} from './run-shared'

interface CollectedItem {
  key: string
  path: string
  kind: RunNode['kind']
  sid: string
  /**
   * The label discovery can derive without a scan: the opening prompt for a
   * session, the spawn description for a subagent. `buildTree` upgrades it to
   * a recorded title once the scan is available.
   */
  label: string
  /** The session's opening prompt, blank when it had none and for subagents. */
  openingPrompt: string
  meta: ClaudeSubagentMeta | null
}

interface CollectedItems {
  items: CollectedItem[]
  unreadable: number
}

/**
 * Subagent metadata is written separately from the transcript and may not exist
 * yet, so a missing file yields `null`. Other storage failures stay typed. A
 * malformed file also yields `null`, after logging why it was rejected.
 */
const readSubagentMeta = Effect.fn('readSubagentMeta')(function*(path: string) {
  const fs = yield* FileSystem.FileSystem
  return yield* fs.readFileString(path).pipe(
    Effect.flatMap(raw => Result.match(parseClaudeSubagentMetaJson(raw), {
      onSuccess: (meta): Effect.Effect<ClaudeSubagentMeta | null> => Effect.succeed(meta),
      onFailure: error => Effect.logDebug('Failed to parse subagent meta', { path, error }).pipe(
        Effect.as(null),
      ),
    })),
    ignoreNotFound(() => Effect.succeed(null)),
  )
})

export const firstPrompt = Effect.fn('firstPrompt')(function*(path: string) {
  const cache = yield* PromptCache
  return yield* cache.get(path).pipe(ignoreNotFound(() => Effect.succeed('')))
})

export const collect = Effect.fn('collect')(function*(
  projectDirectory: string,
  maxAgeHours: number,
) {
  const fs = yield* FileSystem.FileSystem
  const limiter = yield* FileDiscoveryLimiter
  const now = yield* Clock.currentTimeMillis
  const cutoff = freshnessCutoff(maxAgeHours, now)
  const names = (yield* limiter.withPermit(fs.readDirectory(projectDirectory)).pipe(
    ignoreNotFound(() => Effect.succeed([] as Array<string>)),
  )).sort((a, b) => a.localeCompare(b))

  const [sessionFailures, sessions] = yield* Effect.partition(
    names.filter(name => name.endsWith('.jsonl')),
    name => Effect.gen(function*() {
      const path = join(projectDirectory, name)
      if (!(yield* isFreshFile(path, cutoff))) return Option.none<CollectedItem>()
      const sid = name.slice(0, -'.jsonl'.length)
      const openingPrompt = normalizeSessionLabel(
        yield* limiter.withPermit(firstPrompt(path)),
        '',
      )
      return Option.some({
        key: sid,
        path,
        kind: 'session',
        sid,
        meta: null,
        openingPrompt,
        label: openingPrompt || sid.slice(0, 8),
      } satisfies CollectedItem)
    }),
    { concurrency: FILE_CONCURRENCY },
  )

  const [subagentDirFailures, subagentDirs] = yield* Effect.partition(
    names.filter(name => !name.endsWith('.jsonl')),
    sessionName => Effect.gen(function*() {
      const sessionDirectory = join(projectDirectory, sessionName)
      const info = yield* limiter.withPermit(fs.stat(sessionDirectory))
      if (info.type !== 'Directory') {
        return { items: [], unreadable: 0 } satisfies CollectedItems
      }
      const subagentsDirectory = join(sessionDirectory, 'subagents')
      const subagentNames = yield* limiter.withPermit(fs.readDirectory(subagentsDirectory)).pipe(
        ignoreNotFound(() => Effect.succeed([] as Array<string>)),
      )
      const [subagentFailures, subagents] = yield* Effect.partition(
        subagentNames.filter(name => name.endsWith('.jsonl')).sort((a, b) => a.localeCompare(b)),
        name => Effect.gen(function*() {
          const path = join(subagentsDirectory, name)
          if (!(yield* isFreshFile(path, cutoff))) return Option.none<CollectedItem>()
          const agent = name.slice(0, -'.jsonl'.length)
          const meta = yield* limiter.withPermit(
            readSubagentMeta(join(subagentsDirectory, `${agent}.meta.json`)),
          )
          return Option.some({
            key: `${sessionName}/${agent}`,
            path,
            kind: 'subagent',
            sid: sessionName,
            meta,
            openingPrompt: '',
            label: normalizeSessionLabel(meta?.description || '', agent),
          } satisfies CollectedItem)
        }),
      )
      return {
        items: Arr.getSomes(subagents),
        unreadable: yield* countUnreadable(`collect subagents(${subagentsDirectory})`, subagentFailures),
      } satisfies CollectedItems
    }),
    { concurrency: FILE_CONCURRENCY },
  )

  return {
    items: [
      ...Arr.getSomes(sessions),
      ...subagentDirs.flatMap(result => result.items),
    ],
    unreadable: (yield* countUnreadable(`collect sessions(${projectDirectory})`, sessionFailures))
      + (yield* countUnreadable(`collect subagent directories(${projectDirectory})`, subagentDirFailures))
      + subagentDirs.reduce((total, result) => total + result.unreadable, 0),
  } satisfies CollectedItems
})

export const buildTree = Effect.fn('buildTree')(function*(
  projectDirectory: string,
  hours: number,
) {
  const cache = yield* ScanCache
  const limiter = yield* FileDiscoveryLimiter
  const discovery = yield* collect(projectDirectory, hours)
  const [unreadableScans, readable] = yield* Effect.partition(
    discovery.items,
    item => Effect.map(limiter.withPermit(cache.get(item.path)), scan => ({ item, scan })),
    { concurrency: FILE_CONCURRENCY },
  )
  const items = readable.map(entry => entry.item)
  const scans = readable.map(entry => entry.scan)
  const stats = yield* Effect.forEach(scans, scan => scan.stats, { concurrency: 'unbounded' })
  const costSamples = scans.flatMap((scan, index) =>
    scan.diagnostics().context.map(sample => claudeCostSample(sample, items[index]!.key)),
  )
  const byKey = new Map<string, RunNode>()

  for (const [index, item] of items.entries()) {
    const scan = scans[index]!
    // A subagent's transcript carries the parent session's title records, so
    // only a session takes its title from the scan; a subagent keeps the
    // description it was spawned with.
    const title = item.kind === 'session'
      ? normalizeSessionLabel(scan.title, '')
      : ''
    const node: RunNode = {
      ...stats[index]!,
      source: 'claude',
      sourceDetail: 'Claude Code',
      key: item.key,
      kind: item.kind,
      sid: item.sid,
      label: title || item.label,
      title,
      openingPrompt: item.openingPrompt,
      lastPrompt: item.kind === 'session'
        ? normalizeSessionLabel(scan.lastPrompt, '')
        : '',
      agentType: item.kind === 'subagent' && item.meta
        ? item.meta.agentType
        : '',
      toolUseId: item.meta?.toolUseId || null,
      model: item.meta?.model || '',
      spawnDepth: item.meta?.spawnDepth ?? null,
      parentAgentId: item.meta?.parentAgentId || null,
      stoppedByUser: item.meta?.stoppedByUser === true,
      spawnState: '',
      children: [],
      subAgents: 0,
      subRunning: 0,
      subErrors: 0,
      subTools: 0,
      subFiles: {},
      subLast: null,
      subLive: false,
    }
    byKey.set(node.key, node)
  }

  const owner = new Map<string, string>()
  const spawnState = new Map<string, RunNode['spawnState']>()
  for (const [index, item] of items.entries()) {
    const scan = scans[index]!
    for (const id of scan.spawnIds) {
      owner.set(id, item.key)
      // Background launches resolve their tool call instantly, so an async
      // spawn stays running until its task-notification arrives.
      spawnState.set(id, scan.openTools.has(id) || scan.asyncSpawns.has(id) ? 'running' : 'returned')
    }
  }

  const roots: RunNode[] = []
  for (const node of byKey.values()) {
    node.spawnState = node.toolUseId ? spawnState.get(node.toolUseId) || '' : ''
    settleReturnedAgent(node)
    const parentKey = node.toolUseId ? owner.get(node.toolUseId) : undefined
    const parent = parentKey ? byKey.get(parentKey) : undefined
    if (parent && parent.key !== node.key) parent.children.push(node)
    else if (node.kind === 'subagent' && byKey.has(node.sid)) byKey.get(node.sid)!.children.push(node)
    else roots.push(node)
  }

  for (const root of roots) rollup(root)
  roots.sort(bySubLastDesc)
  return {
    roots,
    byKey,
    // Keyed like the other two sources' trees, so the catalog can attribute a
    // skipped record to the session whose transcript it came from.
    scanByKey: new Map(items.map((item, index) => [item.key, scans[index]!])),
    cwd: scans.find(scan => scan.cwd)?.cwd || '',
    malformed: scans.reduce((total, scan) => total + scan.malformed, 0),
    unreadable: discovery.unreadable
      + (yield* countUnreadable(`buildTree scans(${projectDirectory})`, unreadableScans)),
    costSamples,
  }
})

export const buildTrees = Effect.fn('buildTrees')(function*(
  projectDirectories: ReadonlyArray<string>,
  hours: number,
) {
  return yield* Effect.forEach(
    projectDirectories,
    directory => buildTree(directory, hours),
    { concurrency: FILE_CONCURRENCY },
  )
})

/**
 * A completed Agent tool result is stronger evidence than a recently touched
 * subagent transcript. Without this override, returned workers appear active
 * for the entire mtime-based liveness window.
 */
export function settleReturnedAgent(node: RunNode): RunNode {
  if (node.kind === 'subagent' && node.spawnState === 'returned') {
    node.live = false
    node.current = null
  }
  return node
}

export function rollup(node: RunNode): RunNode {
  let agents = 0
  let running = 0
  let errors = node.errors
  let tools = node.tools
  let last = node.lastTs || ''
  let live = node.live
  const files = Object.fromEntries(node.files.map(file => [file.path, file.ops]))

  for (const child of node.children) {
    rollup(child)
    agents += 1 + child.subAgents
    running += (child.spawnState === 'running' || child.live ? 1 : 0) + child.subRunning
    errors += child.subErrors
    tools += child.subTools
    for (const [path, operations] of Object.entries(child.subFiles)) {
      files[path] = (files[path] || 0) + operations
    }
    last = [last, child.subLast || ''].sort().at(-1) || ''
    live ||= child.subLive
  }

  node.subAgents = agents
  node.subRunning = running
  node.subErrors = errors
  node.subTools = tools
  node.subFiles = files
  node.subLast = last || null
  node.subLive = live
  node.children.sort((a, b) => (a.firstTs || '').localeCompare(b.firstTs || ''))
  return node
}

export function flatten(node: RunNode, depth = 0, output: TimelineLane[] = []): TimelineLane[] {
  output.push({
    key: node.key,
    label: node.label,
    agentType: node.agentType,
    kind: node.kind,
    depth,
    firstTs: node.firstTs,
    lastTs: node.lastTs,
    live: node.live,
    errors: node.errors,
    tools: node.tools,
    spawnState: node.spawnState,
    files: node.files.length,
  })
  for (const child of node.children) flatten(child, depth + 1, output)
  return output
}

export function rootOf(roots: RunNode[], key: string): RunNode | null {
  const contains = (node: RunNode): boolean =>
    node.key === key || node.children.some(child => contains(child))
  return roots.find(root => contains(root)) || null
}

export function runPhases(root: RunNode, limit = 16): Milestone[] {
  const phases: Milestone[] = []
  visitNodes(root, (node) => {
    const who = node.kind === 'subagent' ? node.label : 'main'
    phases.push(...node.milestones.map(milestone => ({ ...milestone, who })))
  })
  const selected = phases.some(phase => phase.strong)
    ? phases.filter(phase => phase.strong)
    : phases
  return selected
    .sort(byTimestamp)
    .slice(-limit)
}

export const runDiagnostics = Effect.fn('runDiagnostics')(function*(
  projectDirectory: string,
  root: RunNode,
) {
  const cache = yield* ScanCache
  const nodes: RunNode[] = []
  visitNodes(root, node => nodes.push(node))

  const scans = yield* Effect.forEach(
    nodes,
    node => Effect.flatMap(pathFor(projectDirectory, node.key), path => cache.get(path)),
    { concurrency: FILE_CONCURRENCY },
  )
  const childByToolId = new Map<string, RunNode>()
  for (const node of nodes) {
    if (node.toolUseId) childByToolId.set(node.toolUseId, node)
  }

  const aggregate = emptyRunDiagnostics()
  const costSamples: ClaudeCostSample[] = []

  for (const [index, node] of nodes.entries()) {
    const scan = scans[index]!
    addParseIssueCounts(aggregate.parse.counts, scan.parseIssues.counts)
    aggregate.parse.skipped += scan.parseIssues.skipped
    const diagnostic = scan.diagnostics()
    costSamples.push(...diagnostic.context.map(sample => claudeCostSample(sample)))
    const who = node.kind === 'session' ? 'Main session' : node.label
    if (node.kind === 'session') aggregate.environment = diagnostic.environment
    else {
      for (const key of Object.keys(aggregate.environment) as Array<keyof SessionEnvironment>) {
        aggregate.environment[key] ||= diagnostic.environment[key]
      }
    }

    mergeScanDiagnostics(aggregate, diagnostic, who, node.key)
    if (node.stoppedByUser) {
      aggregate.incidents.push({
        id: `${node.key}:stopped`,
        severity: 'warning',
        category: 'agent',
        title: 'Agent stopped by user',
        detail: node.label,
        ts: node.lastTs,
        line: node.records,
        who,
        key: node.key,
      })
    }
    aggregate.outcomes.push(...diagnostic.outcomes.map((outcome) => {
      const child = childByToolId.get(outcome.toolUseId)
      return {
        ...outcome,
        ...(child ? { childKey: child.key, label: child.label } : {}),
      }
    }))

    const agentUsage = emptyUsage()
    for (const sample of diagnostic.context) addUsage(agentUsage, sample.usage)
    aggregate.agents.push({
      key: node.key,
      label: who,
      agentType: node.agentType || 'Main session',
      models: [...new Set(diagnostic.context.map(sample => sample.model).filter(Boolean))],
      efforts: [...new Set(diagnostic.context.map(sample => sample.effort).filter(Boolean))],
      usage: agentUsage,
      turns: diagnostic.turns.length,
      turnDurationMs: diagnostic.turns.reduce((total, turn) => total + turn.durationMs, 0),
      compactions: diagnostic.compactions.length,
      branchPoints: diagnostic.causal.branchPoints,
      sidechainRecords: diagnostic.causal.sidechainRecords,
    })
  }

  return {
    ...finishRunDiagnostics(aggregate),
    cost: estimateCosts(costSamples),
  } satisfies RunDiagnostics
})

export function stripNode(node: RunNode): PublicRunNode {
  const { children: _children, subFiles: _subFiles, ...publicNode } = node
  return publicNode
}

/**
 * Map a run key to its transcript path.
 *
 * A run key arrives straight from a query parameter, so this is the boundary
 * that keeps a request inside the project directory. The final `resolve` check
 * is the backstop: whatever the segment rules let through, the result must
 * still live under the project root.
 */
export function pathFor(
  projectDirectory: string,
  key: string,
): Effect.Effect<string, InvalidRunKey> {
  const parts = key.split('/')
  if (parts.length > 2 || parts.some(part => !part || part === '..' || part.includes(sep))) {
    return Effect.fail(new InvalidRunKey({ key }))
  }
  const path = parts.length === 2
    ? join(projectDirectory, parts[0]!, 'subagents', `${parts[1]}.jsonl`)
    : join(projectDirectory, `${parts[0]}.jsonl`)
  const root = `${resolve(projectDirectory)}${sep}`
  if (!resolve(path).startsWith(root)) return Effect.fail(new InvalidRunKey({ key }))
  return Effect.succeed(path)
}
